pkg/cli/tail.go

```diff
@@ -0,0 +1,254 @@
+	"github.com/abcxyz/pkg/cli"
+	"google.golang.org/protobuf/encoding/protojson"
+
+	logging "cloud.google.com/go/logging/apiv2"
+)
+
+// Lumberjack specific log types.
+const logType = `LOG_ID("audit.abcxyz/unspecified") OR ` +
+	`LOG_ID("audit.abcxyz/activity") OR ` +
+	`LOG_ID("audit.abcxyz/data_access") OR ` +
+	`LOG_ID("audit.abcxyz/consent") OR ` +
+	`LOG_ID("audit.abcxyz/system_event")`
+
+// logPuller interface that pulls log entries from cloud logging.
+type logPuller interface {
+	Pull(context.Context, string, int) ([]*loggingpb.LogEntry, error)
+}
+
+var _ cli.Command = (*TailCommand)(nil)
+
+// TailCommand tails and validates(optional) lumberjack logs.
+type TailCommand struct {
+	cli.BaseCommand
+
+	flagResource string
+
+	flagValidate bool
+
+	flagMaxNum int
+
+	flagDuration time.Duration
+
+	flagCustomQuery string
+
+	flagRemoveLumberjackLogType bool
+
+	flagValidateWithAdditionalCheck bool
+
+	// For testing only.
+	testPuller logPuller
+}
+
+func (c *TailCommand) Desc() string {
+	return `Tail lumberjack logs from GCP Cloud logging`
+}
+
+func (c *TailCommand) Help() string {
+	return `
+Usage: {{ COMMAND }} [options]
+
+Tails and validates the latest lumberjack log in the last 24 hours from resource:
+
+      {{ COMMAND }} -resource "project/foo" -validate
+
+Tails the latest lumberjack log filtered by additional custom query:
+
+      {{ COMMAND }} -resource "project/foo" -query "resource.type = \"foo\""
+
+Tails and validates (with additional check) the latest 10 lumberjack log in the last 2 hours from resource:
+
+      {{ COMMAND }} -resource "project/foo" -max-num 10 -duration 2h -validate-with-additional-check
+
+Pulls and validates the latest non-lumberjack log type log:
+
+      {{ COMMAND }} -resource "project/foo" -remove-lumberjack-log-type
+`
+}
+
+func (c *TailCommand) Flags() *cli.FlagSet {
+	set := cli.NewFlagSet()
+
+	// Command options
+	f := set.NewSection("COMMAND OPTIONS")
+
+	f.StringVar(&cli.StringVar{
+		Name:    "resource",
+		Aliases: []string{"r"},
+		Target:  &c.flagResource,
+		Example: `projects/foo`,
+		Usage: `Name of the parent resource from which to retrieve log entries,` +
+			`examples are: projects/[PROJECT_ID], folders/[FOLDER_ID],` +
+			`organizations/[ORGANIZATION_ID], billingAccounts/[BILLING_ACCOUNT_ID]`,
+	})
+
+	f.BoolVar(&cli.BoolVar{
+		Name:    "validate",
+		Aliases: []string{"v"},
+		Target:  &c.flagValidate,
+		Default: false,
+		Usage:   `Turn on for lumberjack log validation`,
+	})
+
+	f.IntVar(&cli.IntVar{
+		Name:    "max-num",
+		Aliases: []string{"n"},
+		Target:  &c.flagMaxNum,
+		Default: 1,
+		Usage:   `Maximum number of most recent logs to validate, default is 1`,
+	})
+
+	f.DurationVar(&cli.DurationVar{
+		Name:    "duration",
+		Aliases: []string{"d"},
+		Target:  &c.flagDuration,
+		Example: "2h",
+		Default: 24 * time.Hour,
+		Usage:   `How far back to search for log entries, default is 24 hours`,
+	})
+
+	f.StringVar(&cli.StringVar{
+		Name:    "query",
```

Comment:
This is a bit misleading. This is really _additional_ filters we append to the query, right? Should it be `additional-filter` and make it clear in the usage doc